design/one-pager-stacks-security-isolation.md

```diff
@@ -0,0 +1,239 @@
+A service account cannot grant RBAC permissions that it doesn't already have itself, so to support the arbitrary permissions needs of all stacks, we are currently using the cluster-admin role.
+
+This is a similar problem to what Helm's Tiller has historically had and we can consider revisiting this in the future.
+One possible solution would be to do more client side processing of stacks (instead of in-cluster processing) using the credentials and permissions of the interactive user,
+which would be similar to [Helm 3's](https://github.com/helm/community/blob/master/helm-v3/000-helm-v3.md) major design change.
+
+However, because we are limiting the [allowed set of resources](#allowed-resource-access) that a stack can have access to, this in-cluster Stack Manager model may be acceptable.
+Furthermore, moving to a client side model may limit the full lifecycle management abilities available from an in-cluster Stack Manager, such as automatic stack upgrades.
+
+A full exploration of this topic is out of scope for this version of the design document.
+
+### Stack Package Author
+
+The author of a stack will need to specify the intended scope of permissions for their stack.
+This should be known at stack build time because the stack's controller(s) will need to know whether to watch resources in a single namespace or across all namespaces.
+
+The `app.yaml` schema should be updated with a new field that captures the stack author's intended permission scope:
+
+```yaml
+# Human readable title of application.
+title: Sample Crossplane Stack
+
+# Scope of permissions needed by the stack once installed in the control plane,
+# current supported values are:
+#
+# - Cluster
+# - Namespace
+permissionScope: Cluster
+```
+
+### Stack Install
+
+Since cluster stacks and namespace stacks have differing scopes of execution in the control plane, it will be wise to allow differing access to request their installation.
+A simple way to approach this would be to create different installation request CRDs for each and then use RBAC to restrict access to them by different roles.
+Currently, we use a single `ExtensionRequest` type, but going forward we should differentiate by stack type:
+
+* `ClusterStackInstall`: requests the installation of a cluster stack
+* `StackInstall`: requests the installation of a namespace stack
+
+With these two separate types, an administrator can allow access to installing cluster stacks to only certain restricted roles, then perhaps be more permissive with access to installing namespaced stacks.
+
+If the install scope type does not match what the stack author has declared within their stack metadata, an error status should be set on the install resource and installation of the stack should not proceed.
```

Comment:
What is the reasoning for this restriction? It sounds as though we may be making a Best Practice a Hard Requirement in this case, which I would be generally not in favor of, for [reasons explained in more detail in this other issue](https://github.com/crossplaneio/crossplane-runtime/issues/9#issuecomment-525559769).

Reply:
This should be a hard requirement because of the way controllers can set up their watches for types they care about.  They declare that they should watching within one namespace or should watch across all namespaces.  So if the Stack Manager gives them a scoping of permissions that does not match the stack author's expectations when they coded the controller, it would crash and/or not be functional.  I think we should enforce permission scope to match the stack author's intent.